core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java

```diff
@@ -470,27 +422,156 @@ private void analyze(TokenStream stream, Analyzer analyzer, String field, Set<St
-                if (TypeAttribute.class.isAssignableFrom(attClass))
-                    return;
-                if (includeAttributes == null || includeAttributes.isEmpty() || includeAttributes.contains(key.toLowerCase(Locale.ROOT))) {
-                    if (value instanceof BytesRef) {
-                        final BytesRef p = (BytesRef) value;
-                        value = p.toString();
-                    }
-                    extendedAttributes.put(key, value);
+        stream.reflectWith((attClass, key, value) -> {
+            if (CharTermAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (PositionIncrementAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (OffsetAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (TypeAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (includeAttributes == null || includeAttributes.isEmpty() || includeAttributes.contains(key.toLowerCase(Locale.ROOT))) {
+                if (value instanceof BytesRef) {
+                    final BytesRef p = (BytesRef) value;
+                    value = p.toString();
                 }
+                extendedAttributes.put(key, value);
             }
         });
 
         return extendedAttributes;
     }
+
+    private static CharFilterFactory[] getCharFilterFactories(AnalyzeRequest request, AnalysisService analysisService, AnalysisRegistry analysisRegistry, Environment environment, CharFilterFactory[] charFilterFactories) throws IOException{
+        if (request.charFilters() != null && request.charFilters().length > 0) {
+            charFilterFactories = new CharFilterFactory[request.charFilters().length];
+            for (int i = 0; i < request.charFilters().length; i++) {
+                final String charFilterName = request.charFilters()[i];
+                // parse custom settings if a value start "{"
+                if (charFilterName.startsWith("{")) {
+                    Settings settings = getCustomSettings(charFilterName);
+                    String charFilterTypeName = settings.get("type");
+                    if (charFilterTypeName == null) {
+                        throw new IllegalArgumentException("char filter type does not defined. [" + charFilterName + "]");
+                    }
+                    AnalysisModule.AnalysisProvider<CharFilterFactory> charFilterFactoryFactory =
+                        analysisRegistry.getCharFilterProvider(charFilterTypeName);
+                    if (charFilterFactoryFactory == null) {
+                        throw new IllegalArgumentException("failed to find global char filter under [" + charFilterTypeName + "]");
+                    }
+                    // Need to set custom "name" of char_filter
+                    charFilterFactories[i] = charFilterFactoryFactory.get(getNaIndexSettings(settings), environment, "_custom_charfilter_[" + i +"]", settings);
```

Comment:
Right. I will replace "custom" to "anonymous" in comment, method name, and message.